lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/TaskMessageObserver.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Org.Apache.REEF.Network.Group.Driver.Impl;
+using Org.Apache.REEF.Network.NetworkService;
+using Org.Apache.REEF.Wake.Remote;
+using Org.Apache.REEF.Wake.Remote.Impl;
+
+namespace Org.Apache.REEF.Network.Group.Task.Impl
+{
+    /// <summary>
+    /// The observer for a Task that multiplexes to the node observers associated with that Task.
+    /// </summary>
+    internal sealed class TaskMessageObserver :
+        IObserver<NsMessage<GeneralGroupCommunicationMessage>>, 
+        IObserver<IRemoteMessage<NsMessage<GeneralGroupCommunicationMessage>>>
+    {
+        private readonly Dictionary<NodeObserverIdentifier, IObserver<NsMessage<GeneralGroupCommunicationMessage>>> _observers =
+            new Dictionary<NodeObserverIdentifier, IObserver<NsMessage<GeneralGroupCommunicationMessage>>>();
+
+        private readonly StreamingNetworkService<GeneralGroupCommunicationMessage> _networkService;
+        private readonly object _registrationLock = new object();
+        private bool _hasRegistered = false;
+        private volatile NsMessage<GeneralGroupCommunicationMessage> _registrationMessage;
+
+        public TaskMessageObserver(StreamingNetworkService<GeneralGroupCommunicationMessage> networkService)
+        {
+            _networkService = networkService;
+        }
+
+        /// <summary>
+        /// Registers a node associated with the Task.
+        /// </summary>
+        public void RegisterNodeObserver<T>(NodeMessageObserver<T> observer)
+        {
+            _observers.Add(NodeObserverIdentifier.FromObserver(observer), observer);
```

Comment:
Should it have been? I thought nodes in a graph are uniquely identifiable by the (task, group, operatorName, type) tuple. @dkm2110 can you confirm? I have not yet run into problems with any tests.